src/main/scala/controllers/ActionHandler.scala

```diff
@@ -5,35 +5,54 @@ import javax.inject.Inject
 import cats.data.OptionT
 import cats.instances.future._
 import forms.Field
+import forms.validation.{CostItem, CostItemValues}
 import models._
-import play.api.libs.json.JsObject
+import play.api.libs.json.{JsArray, JsDefined, JsObject, Json}
 import play.api.mvc.Result
 import play.api.mvc.Results._
 import services.{ApplicationFormOps, ApplicationOps, OpportunityOps}
 
 import scala.concurrent.{ExecutionContext, Future}
 
-class ActionHandler @Inject()(applications: ApplicationOps, applicationForms: ApplicationFormOps, opportunities: OpportunityOps)(implicit ec: ExecutionContext) {
+class ActionHandler @Inject()(applications: ApplicationOps, applicationForms: ApplicationFormOps, opportunities: OpportunityOps)(implicit ec: ExecutionContext)
+  extends ApplicationResults {
 
   import ApplicationData._
   import FieldCheckHelpers._
 
   def doSave(id: ApplicationId, sectionNumber: Int, fieldValues: JsObject): Future[Result] = {
-    allValuesEmpty(fieldValues) match {
-      case true => applications.deleteSection(id, sectionNumber).map { _ =>
-        Redirect(routes.ApplicationController.show(id))
-      }
-      case false => applications.saveSection(id, sectionNumber, fieldValues).map { _ =>
-        Redirect(routes.ApplicationController.show(id))
-      }
+    sectionTypeFor(sectionNumber) match {
+      case VanillaSection =>
+        allValuesEmpty(fieldValues) match {
+          case true => applications.deleteSection(id, sectionNumber).map { _ =>
```

Comment:
`saveSection` and `deleteSection` return `Future[Unit]` so there's currently nothing we can check errors on. The happy path is to return to the overview page. The Future will carry exceptions up on the sad path, and at some point we'll be bolstering the http filter to report them more cleanly than it currently does.